hotel/models/hotel_folio.py

```diff
@@ -0,0 +1,825 @@
+            ("manual", "On Check In"),
+            ("picking", "On Checkout"),
+        ],
+        default="manual",
+        help="Hotel policy for payment that "
+        "either the guest has to payment at "
+        "booking time or check-in "
+        "check-out time.",
+    )
+    duration = fields.Float(
+        "Duration in Days",
+        help="Number of days which will automatically "
+        "count from the check-in and check-out date. ",
+    )
+    hotel_invoice_id = fields.Many2one("account.move", "Invoice", copy=False)
+    duration_dummy = fields.Float()
+
+    @api.constrains("room_line_ids")
+    def _check_duplicate_folio_room_line(self):
+        """
+        This method is used to validate the room_lines.
+        ------------------------------------------------
+        @param self: object pointer
+        @return: raise warning depending on the validation
+        """
+        for rec in self:
+            for product in rec.room_line_ids.mapped("product_id"):
+                for line in rec.room_line_ids.filtered(
+                    lambda x, product=product: x.product_id == product
+                ):
+                    record = rec.room_line_ids.search(
+                        [
+                            ("product_id", "=", product.id),
+                            ("folio_id", "=", rec.id),
+                            ("id", "!=", line.id),
+                            ("checkin_date", ">=", line.checkin_date),
+                            ("checkout_date", "<=", line.checkout_date),
+                        ]
+                    )
+                    if record:
+                        raise ValidationError(
+                            _(
+                                """Room Duplicate Exceeded!, """
+                                """You Cannot Take Same %s Room Twice!"""
+                            )
+                            % (product.name)
+                        )
+
+    def _update_folio_line(self, folio_id):
+        folio_room_line_obj = self.env["folio.room.line"]
+        hotel_room_obj = self.env["hotel.room"]
+        for rec in folio_id:
+            for room_rec in rec.room_line_ids:
+                room = hotel_room_obj.search(
+                    [("product_id", "=", room_rec.product_id.id)]
+                )
+                room.write({"isroom": False})
+                vals = {
+                    "room_id": room.id,
+                    "check_in": rec.checkin_date,
+                    "check_out": rec.checkout_date,
+                    "folio_id": rec.id,
+                }
+                folio_room_line_obj.create(vals)
+
+    @api.model
+    def create(self, vals):
+        """
+        Overrides orm create method.
+        @param self: The object pointer
+        @param vals: dictionary of fields value.
+        @return: new record set for hotel folio.
+        """
+        if not "service_line_ids" and "folio_id" in vals:
```

Comment:
> 

This change completely alters the behavior. The original precedence would be like this:
`(not "service_line_ids") and ("folio_id" in vals)`
which would always evaluate to False (because of `not "service_line_ids"`)

If this fixes incorrect behavior shouldn't that rather be a PR with fixing that in an existing branch that would be propagated to other branches and possibly cherry-picked in this PR?

However the code does not make much sense. This is inside _create_ method of _hotel.folio_ model and there is no such attribute as _folio_id_ in that model. So what is actually the intention of that code? Was it to check presence of _folio_id_ in _service_line_ids_ ? But then see remarks regarding _create_ method signature.

Also the code after the `else` is pretty funny. It only handles the situation when `vals` is empty. If it would not be it will end up with nasty exception of using `folio_id` before initialization.

Also the whole method signature:
```
    @api.model
    def create(self, vals):
        """
        Overrides orm create method.
        @param self: The object pointer
        @param vals: dictionary of fields value.
        @return: new record set for hotel folio.
        """
```

Where is this coming from? The _create_ method had _@api.model_ in 11.0 Since 12.0 it is something like:
```
    @api.model_create_multi
    def create(self, vals_list):
```

When _@api.model_ or _@api.model_create_multi_ is used, the parameter _self_ has no meaning at all. Certainly not  a pointer to object. The _@api.model_ on _create_ method will end up being _@api.model_create_single_ which in turn will result in logger warning of not overriding the _create_ method in batch - see https://github.com/odoo/odoo/blob/f01b55636fecf2193ea4bd7551b5daef5bfbb12a/odoo/api.py#L369-L379 and https://github.com/odoo/odoo/blob/f01b55636fecf2193ea4bd7551b5daef5bfbb12a/odoo/api.py#L397-L404